ocaml/xcp-rrdd/bin/rrdd/rrdd_server.ml

```diff
@@ -491,6 +491,33 @@ let dump_host_dss_to_file (file : string) : unit =
       flush oc
   )
 
+let dump_vm_dss_to_file (file : string) : unit =
+  let generate_json rrd =
+    match rrd with
+    | None ->
+        `Assoc []
+    | Some rrdi ->
+        `Assoc
+          (Rrd.ds_names rrdi.rrd
+          |> List.map (fun ds_name ->
+                 (ds_name, convert_value (query ds_name rrdi))
+             )
+          )
+  in
+  let json =
+    let assoc_list =
+      Hashtbl.fold
+        (fun uuid rrd_info acc -> (uuid, generate_json (Some rrd_info)) :: acc)
+        vm_rrds []
+    in
+    `Assoc assoc_list
+  in
+  Xapi_stdext_unix.Unixext.atomic_write_to_file file 0o644 (fun fd ->
+      let oc = Unix.out_channel_of_descr fd in
+      Yojson.Basic.to_channel ~std:true oc json ;
+      flush oc
+  )
```

Comment:
70% code is duplicated to `dump_host_dss_to_file`, I suggest refactor them if possible.

Reply:
I have refacoter the code, I believe there is no more duplicated process now.